File: Swift/Leetcode.playground/Sources/26_Design/LC_146_LRUCache.swift
class DLinkedList : NSObject {
    public var key: Int
    public var val: Int
    public var pre: DLinkedList?
    public var next: DLinkedList?
    public init(_ val: Int,_ key: Int) {
        self.val = val
        self.key = key
    }
    
    override init() {
        self.key = 0
        self.val = 0
    }
}

class LRUCache: NSObject {
    var cache = [Int : DLinkedList]()
    var capacity : Int
    var head : DLinkedList?
    var tail : DLinkedList?
    
    init(_ capacity: Int) {
        self.capacity = capacity
        self.head = DLinkedList()
        self.tail = DLinkedList()
        self.head?.next = self.tail
        self.tail?.pre = self.head
    }
    
    func get(_ key: Int) -> Int {
        guard let node = self.cache[key] else{
            return -1
        }

        self.moveToHead(node)
        return node.val 
    }
    
    func put(_ key: Int, _ value: Int) {
        guard let node = self.cache[key] else {
            let node = DLinkedList(value, key)
            self.cache[key] = node
            self.addToHead(node)
  
            if self.cache.count > self.capacity {
                let node = self.removeTail()
                self.cache.removeValue(forKey: node!.key)
            }
            return
        }

        node.val = value
        self.moveToHead(node)
    }
    
    func addToHead(_ node: DLinkedList) {
        node.pre = head
        node.next = head?.next
        head?.next?.pre = node
        head?.next = node;
    }
    
    func moveToHead(_ node: DLinkedList) {
        self.removeNode(node)
        self.addToHead(node)
    }
    
    func removeNode(_ node: DLinkedList) {
        node.pre?.next = node.next;
        node.next?.pre = node.pre;
    }
    
    func removeTail() -> DLinkedList? {
        let node = self.tail?.pre
        self.removeNode(node!)
        return node
    }
}

/**
 * Your LRUCache object will be instantiated and called as such:
 * let obj = LRUCache(capacity)
 * let ret_1: Int = obj.get(key)
 * obj.put(key, value)
 */

File: Java/26_Design/LC_146_LRUCache.java

class DLinkedList {
    int key;
    int val;
    public DLinkedList pre;
    public DLinkedList next;

    public DLinkedList() {
    };

    public DLinkedList(int key, int val) {
        this.key = key;
        this.val = val;
    };
}

class LRUCache {

    DLinkedList head;
    DLinkedList tail;
    HashMap<Integer, DLinkedList> map;
    int max;
    int count;

    public LRUCache(int capacity) {
        max = capacity;
        count = 0;

        head = new DLinkedList();
        tail = new DLinkedList();

        head.next = tail;
        tail.pre = head;

        map = new HashMap<>();
    }

    public int get(int key) {
        if (map.containsKey(key)) {
            DLinkedList node = map.get(key);
            moveToHead(node);
            return node.val;
        }

        return -1;
    }

    public void put(int key, int value) {
        if (map.containsKey(key)) {
            DLinkedList node = map.get(key);
            node.val = value;
            moveToHead(node);
        } else {
            if (count == max) {
                DLinkedList lNode = removeTail();
                map.remove(lNode.key);
                count--;
            }
            DLinkedList node = new DLinkedList(key, value);
            addToHead(node);
            map.put(key, node);
            count++;
        }
    }

    private void moveToHead(DLinkedList node) {
        removeNode(node);
        addToHead(node);
    }

    private void addToHead(DLinkedList node) {
        DLinkedList hNext = head.next;
        node.next = hNext;
        hNext.pre = node;

        head.next = node;
        node.pre = head;
    }

    private void removeNode(DLinkedList node) {
        DLinkedList pre = node.pre;
        DLinkedList next = node.next;

        pre.next = next;
        next.pre = pre;
    }

    private DLinkedList removeTail() {
        DLinkedList node = tail.pre;
        removeNode(node);
        return node;
    }
}

/**
 * Your LRUCache object will be instantiated and called as such:
 * LRUCache obj = new LRUCache(capacity);
 * int param_1 = obj.get(key);
 * obj.put(key,value);
 */

public class LC_146_LRUCache {

}
